pil-analyzer/src/traits_resolver.rs

```diff
@@ -0,0 +1,131 @@
+    ) -> Self {
+        Self {
+            definitions,
+            identities,
+        }
+    }
+
+    pub fn resolve_traits(
+        &'a self,
+        trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
+        let mut result: HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> = HashMap::new();
+
+        let resolve_references = |expr: &Expression| {
+            expr.all_children()
+                .filter_map(|expr| {
+                    if let Expression::Reference(
+                        _,
+                        Reference::Poly(
+                            reference @ PolynomialReference {
+                                type_args: Some(_), ..
+                            },
+                        ),
+                    ) = expr
+                    {
+                        self.resolve_reference_trait(reference, trait_impls)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>()
+        };
+
+        for (_, (_, def)) in self.definitions.iter() {
+            if let Some(FunctionValueDefinition::Expression(TypedExpression { e: expr, .. })) = def
+            {
+                for (key, value) in resolve_references(expr) {
+                    result
+                        .entry(key.0)
+                        .or_insert_with(HashMap::new)
+                        .insert(key.1, value);
+                }
+            }
+        }
+
+        for identity in self.identities.iter() {
+            for selector in identity
+                .left
+                .selector
+                .iter()
+                .chain(identity.right.selector.iter())
+            {
+                for (key, value) in resolve_references(selector) {
+                    result
+                        .entry(key.0)
+                        .or_insert_with(HashMap::new)
+                        .insert(key.1, value);
+                }
+            }
+            for (key, value) in resolve_references(identity.left.expressions.as_ref()) {
+                result
+                    .entry(key.0)
+                    .or_insert_with(HashMap::new)
+                    .insert(key.1, value);
+            }
+            for (key, value) in resolve_references(identity.right.expressions.as_ref()) {
+                result
+                    .entry(key.0)
+                    .or_insert_with(HashMap::new)
+                    .insert(key.1, value);
+            }
+        }
+
+        result
+    }
+    fn resolve_reference_trait(
+        &self,
+        reference: &PolynomialReference,
+        trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> Option<SolvedImpl> {
+        let name = reference.name.clone();
+        let (trait_decl_name, trait_fn_name) = name.rsplit_once("::")?;
+        if let Some(impls) = trait_impls.get(trait_decl_name) {
```

Comment:
Please add a test where you call a trait function that has zero implementations.

Reply:
This type of error is caught during #1601

Reply:
Are you sure? I think the overlapping can only detect if there are multiple matching implementations.

Reply:
Yes, but since overlapping happens before type checking, it is responsible for catching most of these errors. 

Anyway, if you consider that it is also necessary here, I can add a test for this case and the necessary checks. No problem.